apps/src/lab2/views/components/editor/collaborativeEditorExtension.ts

```diff
@@ -0,0 +1,231 @@
+      const updatesJSON = updates.toJSON();
+      this.channel.send(Action.PUSH_UPDATES, updatesJSON);
+    }
+
+    receiveUpdates(updates: Updates) {
+      console.log('dispatch updates for version', updates.version, updates);
+      this.view.dispatch(receiveUpdates(this.view.state, updates.updates));
+      this.currentVersion = updates.version;
+    }
+
+    onReceive(action: Action, data: any) {
+      if (action === Action.PUSH_UPDATES) {
+        this.receiveUpdates(Updates.fromJSON(data));
+      } else if (action === Action.PULL_UPDATES) {
+        this.receiveUpdates(Updates.fromJSON(data));
+      } else {
+        console.error(`receive(${action}): unknown action`, data);
+      }
+    }
+
+    onConnect() {
+      this.connected = true;
+      this.sendPullUpdates();
+    }
+
+    onDisconnect() {
+      this.connected = false;
+    }
+
+    update(update: ViewUpdate) {
+      if (update.docChanged) this.push();
+    }
+
+    async push() {
+      if (!this.connected || this.pushInProgress) return;
+      const updates = sendableUpdates(this.view.state);
+      if (!updates.length) return;
+
+      this.pushInProgress = true;
+      const version = getSyncedVersion(this.view.state);
+
+      // FIXME: does the web worker version block on this.sendPushUpdates()?
+      // I feel like it must, because otherwise sendableUpdates() contains
+      // the same changes, and the server loses sync from repeatedly transmitting
+      // the same version. We can't directly block on ActionCable, but we could
+      // probably figure out how to make it return a promise that blocks on the
+      // next broadcast received of `push_updates` with the same version sent here.
+      //
+      // The effect of this difference is that if push is ever called while
+      // waiting for the previous sendPushUpdates to be received (and thereby
+      // this.view.dispatch(receiveUpdates(this.view.state, updates.updates))
+      // will be called, which means that update won't show up in sendableUpdates()
+      // anymore), then sync will be totally lost, versions will diverge, and
+      // everything will freak out and usually break. i.e. typing fast-ish
+      // breaks everything and I think this section is to blame.
+      this.sendPushUpdates(new Updates(version, updates));
+      this.pushInProgress = false;
+
+      // Regardless of whether the push failed or new updates came in
+      // while it was running, try again if there's updates remaining
+      if (sendableUpdates(this.view.state).length)
+        setTimeout(() => this.push(), 100);
+    }
+
+    destroy() {
+      this.channel.unsubscribe();
+      this.done = true;
+    }
+  }
+
+  // FIXME: we need to seed our doc and crucial startVersion using Action.GET_DOC
+  // Awkward: how do we block on waiting for the initial version from Action.GET_DOC?
+  // any chance CodeMirror extension creation functions can be async?
+  const startVersion = 0;
```

Comment:
TODO: initiate a get_doc here

Reply:
we're initiating a get_doc in onConnect(), now the question is how to get that data down here, and have everything wait for it....